### PR TITLE
Added `Empirical` and `Implicit` models and distributions.

### PR DESCRIPTION
**Pull request by botev**
Need to add tests, however, it is still worth to review.
Resolving #70

### CODE REVIEW

File: examples/variational_autoencoders/vae.py [hunk 1]
```diff
@@ -17,25 +17,28 @@
 
 
 @zs.reuse('model')
-def vae(observed, n, x_dim, z_dim, n_particles):
+def vae(observed, x_dim, z_dim, n_z_per_x):
     with zs.BayesianNet(observed=observed) as model:
-        z_mean = tf.zeros([n, z_dim])
-        z = zs.Normal('z', z_mean, std=1., group_ndims=1, n_samples=n_particles)
+        z_mean = tf.zeros([1, z_dim])
```
Comment (thjashin) [inline examples/variational_autoencoders/vae.py]: If 1 is used here, all x in the batch are sharing the same `n_z_per_x` samples of `z`?

File: examples/variational_autoencoders/vae.py [hunk 2]
```diff
@@ -17,25 +17,28 @@
...
+        z = zs.Normal('z', z_mean, std=1., group_ndims=1, n_samples=n_z_per_x)
         lx_z = tf.layers.dense(z, 500, activation=tf.nn.relu)
         lx_z = tf.layers.dense(lx_z, 500, activation=tf.nn.relu)
         x_logits = tf.layers.dense(lx_z, x_dim)
-        x = zs.Bernoulli('x', x_logits, group_ndims=1)
-    return model, x_logits
+        x_mean = zs.Implicit("x_mean", x_logits, group_ndims=1)
```
Comment (thjashin) [inline examples/variational_autoencoders/vae.py]: If called `x_mean`, then `tf.sigmoid(x_logits)` is better.

File: examples/variational_autoencoders/vae.py [hunk 3]
```diff
@@ -17,25 +17,28 @@
...
+        x = zs.Bernoulli('x', x_mean, group_ndims=1)
+    return model
 
 
-def q_net(x, z_dim, n_particles):
-    with zs.BayesianNet() as variational:
+@zs.reuse("variational")
```
Comment (thjashin) [inline examples/variational_autoencoders/vae.py]: Since there is only one usage. This can be removed.

File: examples/variational_autoencoders/vae.py (hunk 1 above)
Comment (botev) [reply examples/variational_autoencoders/vae.py]: Ah yes, that is true ... hmm, I can also a `n` than and have both fair enough.

File: examples/variational_autoencoders/vae.py (hunk 2 above)
Comment (botev) [reply examples/variational_autoencoders/vae.py]: Yes, I agree I'll change that as well.

**Comment by botev** [comment]
@thjashin What kind of tests do we want to have for the `Empirical` and `Implicit` distribution?

**Comment by thjashin** [comment]
@miskcoo for ideas.

File: zhusuan/distributions/special.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                 **kwargs):
+        try:
+            self.static_shape = tf.TensorShape(shape[0])
+        except Exception:
+            self.static_shape = tf.TensorShape(tf.Dimension(None))
+        for s in shape[1:]:
+            try:
+                self.static_shape = self.static_shape.concatenate(tf.TensorShape(s))
+            except Exception:
+                self.static_shape = self.static_shape.concatenate(tf.Dimension(None))
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: I guess there should be some internal TF implementation of this feature. From list or TensorShape to TensorShape.

File: zhusuan/distributions/special.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+]
+
+
+class Empirical(Distribution):
+    """
+    The class of Empirical distribution.
+    See :class:`~zhusuan.distributions.base.Empirical` for details.
+
+    :param shape: A list or tuple describing the `batch_shape` of the distribution.
+        The entries of the list can either be int, Dimension or Tensor.
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: Should the list allow Tensor?

File: zhusuan/distributions/special.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        super(Empirical, self).__init__(
+            dtype=dtype,
+            param_dtype=None,
+            is_continuous=is_continuous,
+            is_reparameterized=False,
+            group_ndims=group_ndims,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.constant([], dtype=tf.int32)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: Empirical distributions may not be univariate.

File: zhusuan/distributions/special.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        super(Implicit, self).__init__(
+            dtype=implicit.dtype,
+            param_dtype=implicit.dtype,
+            is_continuous=implicit.dtype.is_floating,
+            group_ndims=group_ndims,
+            is_reparameterized=False,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.constant([], dtype=tf.int32)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: May not be univariate also. But I'm not sure how to deal with this though. Maybe an input argument `value_ndims`?

File: zhusuan/distributions/special.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    def _get_value_shape(self):
+        return tf.TensorShape([])
+
+    def _batch_shape(self):
+        return tf.shape(self.implicit)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: If `value_shape` can be defined by users, this may be inferred.

File: zhusuan/distributions/special.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    def _get_batch_shape(self):
+        return self.implicit.get_shape()
+
+    def _sample(self, n_samples):
+        implicit = tf.expand_dims(self.implicit, 0)
+        return tf.tile(implicit, [n_samples] + [1] * self.implicit.shape.ndims)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: Maybe we shouldn't allow `n_samples` to be used for Implicit distributions? This `tile` seems strange.

**Comment by thjashin** [comment]
Also @ssydasheng for a look at the `Implicit` distribution.

File: zhusuan/distributions/special.py (hunk 6 above)
Comment (botev) [reply zhusuan/distributions/special.py]: I guess this would require an optional argument to be passed for this

File: zhusuan/distributions/special.py (hunk 9 above)
Comment (botev) [reply zhusuan/distributions/special.py]: Sure, I was considering that as an option as well.

File: zhusuan/distributions/special.py (hunk 5 above)
Comment (botev) [reply zhusuan/distributions/special.py]: The issue, in fact, for using tensor is the one listed above. If the shape is a tensor `tf.TensorShape` does not work with it so I need to check for the exception.

**Comment by botev** [comment]
@thjashin Should the empirical distribution have inputs `shape, value_shape` or `batch_shape, value_shape` as both are possible?

**Comment by thjashin** [comment]
@botev Sorry I missed this comment. I prefer `batch_shape` and `value_shape`.

**Comment by botev** [comment]
@thjashin The last issue remaining is whether to allow tensors to be passed as `batch_shape` and `value_shape`. The issue currently with passing tensors is that `tf.TensorShape` does not automatically convert them to `None` so I will have to do the hack myself with `try - catch` otherwise if tensors are not allowed everything will work fine (e.g. list of ints and Nones).

**Comment by thjashin** [comment]
I think just static shapes are enough. Since this distribution should always be observed, there is no case that its dynamic shapes will be used. For static shape, we may use it to check shape compatibility with observations. Though this is not currently implemented in `StochasticTensor`.

**Comment by botev** [comment]
I've updated the code as discussed. Not sure what tests to add though?

**Comment by thjashin** [comment]
Basically it should be some checks that cover all lines of the added code. Specifically, you need checks on the input verification, and checks on shapes (batch shape, value shape, sample shape, and log_prob shape), and checks of the numerical results of log_probs/probs. You can refer to tests of `Normal` and `OnehotCategorical` for examples on univariate/multivariate distribution.

**Comment by botev** [comment]
Sorry, I did not update on this but I was at NIPS and got a bit busy before the holidays. 
I'll try making an update on this next week.

**Comment by thjashin** [comment]
That's fine. Sad I didn't go to nips this year. Or we could have chat.

**Comment by botev** [comment]
I've added some tests for the distributions. Could you take a look if more are needed.

**Comment by botev** [comment]
This should be ready to merge now.

File: zhusuan/distributions/special.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            raise ValueError("ImplicitDistribution does not accept `n_samples` argument.")
+        return tf.expand_dims(self.implicit, 0)
+
+    def _log_prob(self, given):
+        return tf.log(self.prob(given))
+
+    def _prob(self, given):
+        given = tf.cast(given, self.param_dtype)
+        given, implicit = maybe_explicit_broadcast(given, self.implicit, 'given', 'implicit')
+        prob = tf.cast(tf.equal(given, implicit), tf.float32)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: The two casts are unnecessary. Since given will be checked to be compatible with self.dtype.

File: zhusuan/distributions/special.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        if n_samples is not None and n_samples != 1:
+            raise ValueError("ImplicitDistribution does not accept `n_samples` argument.")
+        return tf.expand_dims(self.implicit, 0)
+
+    def _log_prob(self, given):
+        return tf.log(self.prob(given))
+
+    def _prob(self, given):
+        given = tf.cast(given, self.param_dtype)
+        given, implicit = maybe_explicit_broadcast(given, self.implicit, 'given', 'implicit')
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: No need to explicit broadcast.

**Comment by thjashin** [comment]
@botev Further comments before merging.

File: zhusuan/distributions/special.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+                 group_ndims=0,
+                 is_continuous=None,
+                 **kwargs):
+        self.explicit_batch_shape = tf.TensorShape(batch_shape)
+
+        if dtype is None:
+            dtype = tf.float32
+        assert_same_float_and_int_dtype([], dtype)
+
+        if value_shape is None:
```
[the remark this answers is not in the document]
Comment (botev) [reply zhusuan/distributions/special.py]: Shouldn't, in that case, the default value_shape be `[]` as I think that is the most standard case.

File: zhusuan/distributions/special.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+        super(Empirical, self).__init__(
+            dtype=dtype,
+            param_dtype=None,
+            is_continuous=is_continuous,
+            is_reparameterized=False,
+            group_ndims=group_ndims,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.convert_to_tensor(self.explicit_value_shape, tf.int32)
```
[the remark this answers is not in the document]
Comment (botev) [reply zhusuan/distributions/special.py]: Yeah, that sounds like an option. I was not too sure what to do with it either.

File: zhusuan/distributions/special.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (thjashin) [reply zhusuan/distributions/special.py]: For an arbitrary stochastic generator, if the samples are univariate, the value_shape is `[]`, or else they are multivariate, the value_shape has one more rank. So the value_shape is actually determined if you know about the semantics of the samples. In that sense the default should be the case that you don't know this or it doesn't matter. In that sense, a better default should be <unknown> (`TensorShape(None)`). Does this make sense to you?

**Comment by botev** [comment]
@thjashin I think I've addressed all of the recommendation so just a final review needed.

File: tests/model/test_stochastic.py [hunk 14]
```diff
@@ -315,3 +315,44 @@ def test_ExpConcrete(self):
...
+            b = Implicit('b', a, value_shape=[])
+
+        sample_ops = set(get_backward_ops(b.tensor))
+        for i in [mean, logstd, n_samples]:
+                self.assertTrue(i.op in sample_ops)
+
+        self.assertEqual(a.get_shape().as_list(), b.get_shape().as_list())
+        (a_value, ),  (b_value, ) = model.query(['a', 'b'], outputs=True)
+        # The ops are Squeeze(ExpandDims(a))
+        self.assertEqual(a_value, b_value.op.inputs[0].op.inputs[0])
```
Comment (thjashin) [inline tests/model/test_stochastic.py]: The order of inputs seems to be undefined behaviors. Better to remove this.

**Comment by thjashin** [comment]
Almost good. Only minor comments.

File: zhusuan/distributions/special.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    """
+
+    def __init__(self,
+                 implicit,
```
[the remark this answers is not in the document]
Comment (botev) [reply zhusuan/distributions/special.py]: Inputs does not seem right. Samples seems a bit better. What do you think about `variables` or `transformed_variables`?

[the remark this answers is not in the document]
Comment (thjashin) [reply zhusuan/distributions/special.py]: But variables are reserved words in TF. How about `simulations`? Doesn't seem to be better than `samples` though.

[the remark this answers is not in the document]
Comment (botev) [reply zhusuan/distributions/special.py]: `transformed_samples`?

File: zhusuan/distributions/special.py [hunk 16]
```diff
@@ -152,8 +150,11 @@ def _log_prob(self, given):
         return tf.log(self.prob(given))
 
     def _prob(self, given):
-        prob = tf.cast(tf.equal(given, self.implicit), tf.float32)
+        prob = tf.equal(given, self.implicit)
+        print(self.param_dtype)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: print

File: zhusuan/distributions/special.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (thjashin) [reply zhusuan/distributions/special.py]: A bit too long... Let it be `samples` for now. I'll merge.

**Comment by botev** [comment]
Should be finally done.

**Comment by thjashin** [comment]
Seems that you committed some images.

**Comment by botev** [comment]
Sorry about that, fixed it now.

**Comment by thjashin** [comment]
Nice work! Thank you for contributing.
